doc/aaa/TACACS+ Design.md

```diff
@@ -0,0 +1,521 @@
+
+ - Following event will be accounted:
+ 	- User login to SONiC host.
+ 	- User logout.
+ 	- User run command on host: 
+		- Command start run.
+		- Command finish.
+
+- User command in Docker will not be accounted.
+    - User command in docker actually run by docker service, so we can't identify if command are run by user or system service.
+	
+- Failover:
+	- Use syslog as backup when remote TACACS not not accessible from SONiC.
+
+
+## 1.4 User script
+ - User can create and run their own script. 
+ - If user create a script but TACACS+ service side not have configuration to allow user run this script, user script will be blocked by authorization.
+
+## 1.5 Docker support
+ - Docker exec command will be covered by Authorization and Accounting. 
+ - Any command run inside docker container will not covered by Authorization and Accounting.
+
+## 1.6 Multiple TACACS server
+ - Support config multiple TACACS server.
+ - When a server not accessible, will try next server as backup.
+ - When all server not accessible from SONiC, use native failover solution.
+
+# 2 Configuration and Management Requirements
+## 2.1 SONiC CLI
+ - Enable/Disable TACACS Authorization/Accounting command
+```
+	config aaa authorization {local | tacacs+}
+    config aaa accounting {local | tacacs+}
+    config aaa accounting local
+```
+
+ - Counter command
+```
+    show tacacs counter
+    sonic-clear tacacscounters
+```
+
+## 2.2 Config DB
+ - TACACS AAA are fully configable by config DB.
+
+## 2.3 Counter
+ - Support  AAA counter:
+```
+    show tacacs counter
+    
+        server1: 10.1.1.45
+        Messages sent: 24
+        Messages received: 20
+        Requests accepted: 14
+        Requests rejected: 8
+        Requests timeout: 2
+        Requests retransmitted: 1
+        Bad responses: 1
+```
+
+## 2.4 Syslog
+ - Generate syslog when Authentication/Authorization/Accounting.
+ - When remote TACACS server not accessible from SONiC, use syslog for accounting.
+
+
+# 3 Limitation
+## 3.1 Command size
+ - TACACS protocol limitation: command + parameter size should smaller than 240 byte. The longer than 240 bytes parts will be drop.
+ 	- This limitation is a protocol level, all TACACS implementation have this limittation, include CISCO, ARISTA and Cumulus.
+ 	- Both Authorization and Accounting have this limitation.
+ 	- When user user a command longer than 240 bytes, only commands within 240 bytes will send to TACACS server. which means Accounting may lost some user input. and Authorization check can only partly check user input.
+
+## 3.2 Server count
+ - Max TACACS server  count was hardcoded, default count is 8.
+
+## 3.1 Local authorization
+ - Operation system limitation: SONiC based on linux system, so permission to execute local command are managed by Linux file permission control. This means TACACS+ authorization can't config to disable 'local' authorization, and local authorization must be last authorization in authorization method list.
```

Comment:
>can't config to disable 'local' authorization [](http://example.com/codeflow?start=179&length=45)

If tacacs server is not reachable during authorization, we have option to fail the authorization, so local auth is disabled.

Reply:
Fixed.

Reply:
My points is that you have 2 options
1. tacacs + local
2. tacacs only, no local
And this is not a "Operation system limitation".

Reply:
Here is a example of this issue:
           user name: domainuser
           When user login, tacacs return privilege level 0 so in SONiC this user is a RO user, then this user on sonic only have Linux permission to run RO commands. 
           But in tacacs server side, the 'config' command was config as allow user run.

Then when user trying to run the 'config' command, tacacs authorization succeeded, but user still can't run this command because user not have Linux permission.

In EOS system, the behavior is different:
          We add domainuser as local RO user.
          In tacacs server config this user can run RW command.
         Then we user run a RW command, user can run it.

The reason is the EOS system can bypass the local permission check, but SONiC is a Linux system, and currently we using Linux permission for 'local authorization', so tacacs authorization can't bypass the Linux permission check.

Reply:
I see your point!
Actually I am considering another config "tacacs only, no fallback to local", which means tacacs server unreachable, will stop the already-login user to run anything except `logout` or `exit`. Then the user need to re-login with local account, local authentication and local authorization.